### ISSUE TITLE
Separation between issue and task is not clear nor practical

### ISSUE THREAD

**Issue by allan-simon** [taigaio/taiga-front#790]
The user documentation does not explain clearly what is the goal of both, but some things that I've seen:

(disclaimer: I'm very new to agile method so maybe some of my confusion is only because i'm not clear about scrum methods)

As a manager the most practical view for me to see at one time where my team is the taskboard
However in the taskboard I can't see the issues,  so now i have two places to look for

Which create confusion for me,
1. if i use only issues, I don't have the taskboard and related things
2. if I use only user stories and task, I don't have the easy way to report quickly a bug  （I need to create a user story , then split it into a task to have it in the taskboard)
3. If I use both , i now have two places to check progress of my team

All of that for me could be solved by simply being able to see issues in the taskboard

**Comment by Noogic** [comment]
I think issues are an early stage of a user stories and a way to communicate about the project, it's needs, it's possible features and it's bugs without intervening the actual work and sprints.

It has it's advantages because you don't have to worry about writting full user stories or what is already in the backlog or a sprint.

An issue usually evolves into a user story which also evolves (or breaks) into tasks. So every task is related to a user story which in the end will provide new and *measured* value while an issue is something more abstract and can be a bug, the start of a new user story or just a question.

**Comment by allan-simon** [comment]
but
   * you can't categorize a user story as a bug  (like how would you do for a bug "the title has a typo in spanish version" ?) 
   * nor do i want to go the full process just for tracking "add 'use strict'  in our javascript code" 

also as for the backlog, as long as your issues are not associated to a sprint, they will not polluate anything

**Comment by Noogic** [comment]
How to work properly with bugs is something I haven't figured out yet.

I think there are two kinds of bugs: an implementation bug and a design bug.
· An *implementation* bug occurs when you have a user story, it has been completed and yet, something wrong happens. It may be a test or QA fault.
· A *design* bug happens because there is an scenario that has not been taken into account, so there is not a user story or task which prevents this for happening.

- Should a bug's issue evolve into a user story when the bug is from design? Probably yes.
- Should a bug's issue evolve into a user story when the bug is from implementation? Well, you already have a user story, so this is the scenario where I'm not sure how to work with.

In taiga, you can promote a bug issue to a user story, and the user story keeps a reference to the issue, so you can relate everything about the bug in the issue and create all the tasks to solve it in the user story.

The only thing I miss is create a complete attachment between issue and user story so when the user story has tasks *in progress*, the issue changes itself to *in progress* and when the user story has all its tasks in *closed*, asks the user assigned to the user story if wants to close the issue too.

More than that, I've discovered taiga just yesterday. I've been working with agile since last year and still learning about it. So, related to taiga's workflow I need more time to figure out all it's details.

So, in summary, I see issues as a way to log everything related to a project: bugs, questions and desired features, while I see user stories as a quality content to create value to a project, with scenarios and all the needed details to create that value. In order to complete a user story, you break it into tasks.

In your case, I would suggest to create issues. Someone in your team then promote an issue to a user story. In that moment, a comment is placed in your issue, so you will know it has been promoted to a user story. Then, you can go to that user story and use the *watch* button in order to keep a track of it. Finally, when the story is closed, the one who closes it can also close the issue or you can do it yourself since you know the story is closed.

Hope it helps :)

**Comment by Noogic** [comment]
Aaaand, for your last comment "*also as for the backlog, as long as your issues are not associated to a sprint, they will not polluate anything*", I think you can mess the backlog if you create stories/issues without the style/convention about user stories.

You can create a issue which says *I want to have the website in other languages* and this is great because this is a possible feature and now the whole team is aware of it. But, in sprints, features needs to be prioritize the give always as value as possible. In order to do that, besides the feature a user story needs the reason of a feature to be implemented. You can achieve that using the [5 whys technique] (https://en.wikipedia.org/wiki/5_Whys). Also, you should define scenarios in order to avoid bugs, and you can do that using the [Given then when technique] (http://guide.agilealliance.org/guide/gwt.html).

Do you see now the difference (or at least the possible difference, there could be more) between an *issue* and a *user story* ?

**Comment by allan-simon** [comment]
I do see the difference between an issue and a user story, it  was more about tasks  and issue. thanks for the other explanations and links, they are interesting.

But you what you expose

```
In your case, I would suggest to create issues. Someone in your team then promote an issue to a user story. In that moment, a comment is placed in your issue, so you will know it has been promoted to a user story. Then, you can go to that user story and use the watch button in order to keep a track of it. Finally, when the story is closed, the one who closes it can also close the issue or you can do it yourself since you know the story is closed.
```

is sooooo much more complicated than our current redmine workflow for such trivial bug

 1. tester/user report the bug
 2. developer fix an estimate (in point/hours)   
 3. manager/developers decide to associate it to a sprint 

and that's done,  all the information are in one place about the issues, (the link issues) ,  and it does appear automatically in our redmine plugin that provide us a taskboard

also for the backlog, sorry i misexpressed myself, what i had in mind was that

  1. you have your backlog with user story, and only user story
  2. in your list of issues, you can associate them to a sprint (and hence see them in the taskboard)
  3. you can put an estimate to an issue 



to summarize

  1. I don't want to do any intellectual gymnastic things to write a user story around a trivial bug or technical debt ( e.g "convert all mixed tab/spaces by only spaces" ) just for the sake of making it fit in an artificial model
  2. I want as a manager to see in one place (the taskboard) what things are going to be done this sprint and their status

**Comment by Noogic** [comment]
Sorry I can't be of more help, I'm also trying to fully understand Agile and Scrum, which are complicated.

You also have said being new into Agile, and it can look easy at first and overwhelming when you try to apply to real life, just as happens with TDD.

In my personal experience, trying to apply a new methodology (or new approach, design, whatever) without understanding why it does what it does can go really bad.

Usually it follows like this:

> My system doesn't work -> I find a new system which works well for other people -> I learn about it and try to implement it -> It fails.

**Why it fails if it seems you've followed all the steps and have done everything like others?**
In my opinion it is because of the details. When we are following all the steps, we are not really aware of the most tiny things and we tend to do as we are used to.

So, when you say "*I don't want to do any intellectual gymnastic things to write a user story around a trivial bug or technical debt ( e.g "convert all mixed tab/spaces by only spaces" ) just for the sake of making it fit in an artificial model*" **maybe** you're wrong (and I say maybe because here is where we need someone with more experience and knowledge).

SCRUM is great, yes, it solves problems we have with other methodologies, yes, but everything comes at its costs. So, you have a workflow which consists in something like this:
> Identify problems and possible features (Issues) -> debate about them (Comments in issues) -> Evolve into a user story (US) and calculate costs/points -> Prioritize user stories (Backlog) -> Create Sprint

If it works but suddenly you have a simple bug or simple task and you want to break the normal workflow because it's too much overwhelming, be sure you understand how everything works and why it works, otherwise you will be paying a cost you don't know. And this, is just my opinion but it comes from experience.

The key in Scrum are the user stories. An user story has not only all the information about the feature to implement but also has all you need to know how much everything costs. And this could be as important as the user story itself. The same in TDD is not about writing test about everything but writing the right test, in Scrum is not about creating user stories but creating the right user stories and is not about sprints but the right sprints.

So maybe you're right and the current way of handling issues and tasks is not the best, but maybe it has its reasons to be like this. Which are these reasons is what we need to find in order to make changes to the usual workflow safely.

**Comment by Noogic** [comment]
Lets put your example in practice:

- convert all mixed tab/spaces by only spaces

It seems a simple thing to do. If you create an issue, you can put this in it just that and it's ok.
What you complain about is having to create also a user story for this once you have created the issue.

So, why can't you just put this issue in the sprint? Is not because of its complexity but because of the costs for the sprint.

Something like this must be converted into a user story in order to all the departments being able to identify the cost for them.

The conversion we are trying to apply will affect to every file with code which also will create conflicts with all the create code since the start of the task. This is a cost the scrum manager must know and take into account. The user story can have several tasks:

- Create the script to replace tabs/spaces to only spaces (which also needs to be tested for each scenario)
- Block any CI for this sprint
- Avoid merges of other features until this has been closed
- Apply this feature to other branches before merge (why?, to avoid conflicts)

It could be easier or more difficult, but you wont know until you create the user story. The user story will give you the real costs and the actual tasks while the **5 whys** methods will give you the importance of this user story.

**Comment by allan-simon** [comment]
@Noogic , sure I understand better the logic in what you're saying :) well let's see if one of the project member or someone with more experience in scrum can bring us more light

but as i was explaining in #756 , even though I agree that what you're explaining is certainly best and ultimately what we should do, but I don't want a tool to impose that on me on day 1,  and i'm already enthusiastic about the tool. 

Now imagine imposing this on my team members which for some of them maybe less enthusiastic, and you're pretty sure the tool will not get adopted because it's just too far a jump.

It's like for python and the pep8 codestyle, you don't jump from "code that has never followed anycode style" to "pep8 compliant code" ,  all these tools permit you to deactivate rules and activate them one by one as your team get used to them. 

Same here, I want the tool to let me use it , regardless on how bad I use it, so that it can win the hearts of people by being a "prettier and less clumsy redmine, but same workflow" so that people forgot about redmine  and then from here you go on "ok this tool permit us to have user stories, backlog and stuff, let's use it" and from here introduce your arguments (that i will certainly make mine on the long run) which are all pretty clear and full of sense. Because then if people find your arguments complicated etc.,  you just stay on "using taiga as a redmine" , and you can retry the "best practices way speech" in some month when it starts to get clearer that following them would have clear advantages.

**Comment by Noogic** [comment]
@allan-simon you are totally right. Trying to implement new methodologies or tools to a team is always difficult. Much people never want to change. Be sure they see why you need to change and why it can't wait. Also see what other people in your role say about introducing agile (or making changes in general) to a consolidated team. And remember, you are not trying to move your people only to taiga, you are also trying to move them to agile. If they don't embrace agile probably wont like taiga.

Finally, if you really need to make this changes to Taiga in order to your team to use it, you depend on the taiga's team, but being an open source project you have many options to get what you want:

1. You can create the feature and make a pull request so you don't have to wait until someone else does.
2. If you create the feature but the taiga's team does not want it (I'll explain later), you can use the self-hosted version and apply the changes in your own server.
3. You can create a plugin/service to automate the process. You wont have issues in the task board, but you can make them to be automatically synced with a task. You can achieve this using the API and webhooks.

>Note: It is probably they don't want a feature just to make easy a transition because any option comes with its own complexity. It's what happens with Jira, for example, and with any atlassian software. It has tons of options and features but that makes it harder to use and learn. Making a sofware simpler keeping the most valuable features and avoiding the *just cool* features has its worth. If this is the case, try with the third option: sync issues with tasks using webhooks and API.

**Comment by allan-simon** [comment]
Sure, being myself involved in quite some opensource project I do know about the "be the change you want to see in the world" :) (I do have made a plugin to plug with ping federate auth system for taiga)  but for your different points 

   1.  before creating the pull request, I would like to be sure it does not against the core principle of the taiga team, hence the issue and comments we're currently writing. If it's ok, then I would happily make a PR.
   2.  yes i could fork, but then I would need to start manually to backport any bug fix, new features etc. so the apparent simple solution become a burden on the longterm. 
   3. unfortunately if i need to resort to this, i will certainly just keep with redmine :/ because "keeping in sync" would be easy if it was a 1 to 1 mapping,  but here it means you need to create a user story with a task for each issue (as the goal is to have it in the taskboard), and now we're polluating the backlog with things that are not user stories as you said, and we're not talking about comments that are posted both way (or on the task associated to the user story)

for your side note, yes i totally agree. that's why I think being able to have issues in taskboard would be the simplest way to solve the problem.
After sure, I do also understand that my request is one among a hundreds of other that will maybe a death by a thousand cut for taiga.

**Comment by Noogic** [comment]
Glad to help. And thank you too for your question and concerns, I'm also interested in that. It seems there is no standard or agreement on how to handle *non user stories*.

In Jira, for example, only user stories can be estimated. Bugs and other types can't. After reading your other issue I've realized the problem is not about if you should be able to put issues directly into the sprint or not, it is about how to handle this issues. Maybe using the 5 whys methods with "*why do you want to put issues directly into a sprint*" give us more information about that.

Looking for more information, I've found this:
http://www.mountaingoatsoftware.com/blog/should-story-points-be-assigned-to-a-bug-fixing-story
http://www.agileforall.com/2010/05/agile-antipattern-sizing-or-estimating-bug-fixes/

So, after this issue and the https://github.com/taigaio/taiga-front/issues/756 I think issues should be allowed to be in taskboard but not be able to have an estimate by default. Not being able to assign points to a bug makes you wonder why and then you make your research and realize the problem about estimating a bug. Then, you can safely assume the risk of doing that, so you go to config and allow issues to be estimated at your own risk.

Surely this will be a hard topic and eventually we will learn more about it. Now we have to wait to know what are the principles of the team and how intend to solve this.

**Comment by allan-simon** [comment]
Yes i would be perfectly fine with a middle ground of not being able to estimate task (or only with a custom field that does mix in with the point system of user stories)

It's a bit late here and I will read your links tomorrow.
